### ISSUE TITLE
WO-004: RNG

### ISSUE THREAD

**Issue by Xocas12** [Xocas12/gosplan-env#9]
> **WO-004** - Phase 1 - Owner **MID-fast** - Difficulty **2/5** - Gate **G0**
> Card: [`workorders/WO-004.md`](https://github.com/Xocas12/gosplan-env/blob/main/workorders/WO-004.md)

## What this is

Every random number the environment ever uses comes out of one function, `draw`. This card writes
its single body: a key-based generator that builds a `SeedSequence` from `(seed_env, crc32(purpose),
*indices)` and returns values from one of four distributions. No stream, no counter, no global
generator.

It exists because the whole design leans on reproducibility properties that only a key-based
construction buys: a trajectory that reproduces regardless of call order, common random numbers
across experimental arms so paired contrasts are valid without variance-reduction tricks, and a
Phase-2 JAX port that agrees with the NumPy environment by construction rather than by luck. A
module-level generator would destroy all three silently.

## Depends on

- WO-001
- WO-002

WO-001 is the formal dependency (PLAN section 12.2); the frozen suite this card must pass is written
by WO-002, so both must have landed before this card is issued.

**Blocked by:** #6 (WO-001), #7 (WO-002)
This card may not be started until every issue above is closed (CONTRACT rule 12 / the loop in CONTRIBUTING.md).

## Read first (exhaustive whitelist)

- [`spec/spec.py`](https://github.com/Xocas12/gosplan-env/blob/main/spec/spec.py) - the frozen `draw`
  signature and the `Purpose` / `Dist` literals.
- [`gosplan/rng.py`](https://github.com/Xocas12/gosplan-env/blob/main/gosplan/rng.py) - **the write
  target.** Module docstring, `PURPOSES`, `Purpose`, `Dist`, `Array`, and the complete `draw`
  contract are already there; only the body is missing.
- [`CONTRACT.md`](https://github.com/Xocas12/gosplan-env/blob/main/CONTRACT.md) - rule 9 above all
  (all environment randomness goes through `draw`; no direct `numpy.random` / `jax.random` in
  `gosplan/env/`; `seed_policy` is separate), plus rules 3 and 12.
- [`pyproject.toml`](https://github.com/Xocas12/gosplan-env/blob/main/pyproject.toml) - ruff
  line-length 100, target py312.
- [`tests/unit/test_rng.py`](https://github.com/Xocas12/gosplan-env/blob/main/tests/unit/test_rng.py) -
  read-only (CONTRACT rule 2); written by WO-002, so it is not in the tree until that card lands.

Read nothing else. CONTRACT rule 12.

## Write only

- `gosplan/rng.py`

Creating, editing or deleting any other file is a CONTRACT rule 12 violation.

## Must pass

- [ ] `tests/unit/test_rng.py` - **T-U6**: `draw` is deterministic in `(seed_env, purpose, indices)`
      and independent of call order; the four distributions `lognormal(mean_log, sigma)`, `normal`,
      `bernoulli`, `categorical` have the stated moments.

```bash
uv run pytest tests/unit/test_rng.py -q
uv run ruff check gosplan/rng.py
uv run ruff format --check gosplan/rng.py
```

## Implementation notes

One body: `draw(seed_env, purpose, *indices, shape, dist, **params) -> Array`. The construction is
fixed by PLAN section 2.15 and the WO-004 notes of PLAN section 12.3, and every reproducibility
property in the design depends on following it exactly:

```
SeedSequence([seed_env, crc32(purpose), *indices])  ->  default_rng(seq)  ->  values
```

with `crc32` being `zlib.crc32(purpose.encode())`. Use `zlib.crc32`, **never** Python's builtin
`hash`: `hash` is salted per process, so a run would not reproduce across processes and CONTRACT
rule 10's manifest would be worthless. Nothing else enters the key - no call counter, no wall clock,
no length of a previous draw, no `shape`.

### Distribution table

Parameter names are part of the contract; callers pass them by keyword.

| dist | parameters | returns |
|---|---|---|
| `lognormal` | `mean_log`, `sigma` | `exp(N(mean_log, sigma**2))`; float array |
| `normal` | `mean`, `sigma` | float array |
| `bernoulli` | `p` | **boolean** array |
| `categorical` | `probs` | **integer** array of category indices |

Return exactly the requested `shape`, with the dtype the table states. `shape` vectorises over the
trailing index: the environment calls, for example,
`draw(seed_env, "yield", t, k, shape=(N,), dist="lognormal", mean_log=..., sigma=...)` for one shock
per enterprise, so the whole `(N,)` vector comes from **one** key rather than from `N` keys.

### Properties the construction buys, all checked by T-U6

- Determinism in `(seed_env, purpose, indices, shape, dist, params)` and **independence of call
  order**, which is what makes the NumPy environment and the Phase-2 JAX port agree by construction
  (the WO-029 parity test) and what lets the period schedule of PLAN section 2.5 be reordered
  without changing a trajectory.
- Common random numbers across arms whenever `seed_env` is shared (PLAN section 4.3), which is what
  makes the paired contrasts valid without variance-reduction tricks.
- Independence of distinct purposes at identical indices.
- Complete separation of `seed_policy`, which drives policy sampling and never appears in a key
  here.

### Errors, raised rather than papered over (all `ValueError`)

- a `purpose` outside `PURPOSES`;
- a `dist` outside `Dist`;
- a missing distribution parameter, or an unknown one. **Do not default `sigma` to 1, do not default
  `mean` to 0, do not accept `scale` as an alias for `sigma`.** A caller that omits a parameter has a
  bug, and defaulting it would hide a silently wrong shock in a trajectory that still reproduces.

### Edge cases

- `numpy.random.SeedSequence` takes non-negative integer entropy. `crc32` is non-negative and `t`,
  `k`, `i` are non-negative by construction. If a caller ever needs a negative index, that is an
  interface question: file an AMBIGUITY REPORT. Do not take an absolute value, do not offset, do not
  hash the tuple.
- `sigma = 0.0` is legitimate and common in Phase 1 (`audit_noise`, `channel_noise`,
  `self_obs_noise` are all `0.0`). It must return exactly `mean` for `normal` and exactly
  `exp(mean_log)` for `lognormal`, with no warning, and it must **still consume a key** so that the
  key layout does not depend on the configuration.
- `probs` for `categorical` arrives as a sequence; it must not be mutated, and it is the caller's
  business whether it sums to 1 - validate that it does and raise if it does not.
- The yield shock of PLAN section 2.6 is `lognormal` with `mean_log = -sigma**2 / 2`, so
  `E[eps] = 1`. That algebra belongs to the **caller** (`gosplan/env/production.py`, WO-005), not
  here; `draw` takes `mean_log` as given. `tests/unit/test_production.py` checks the mean to 1e-3
  over 1e5 draws.

## Forbidden

- Module-level global generators, of any kind: no module-level `Generator`, no cached
  `SeedSequence`, no counter, no memo keyed on anything but the arguments of the call. A global
  reintroduces call-order dependence through the back door and destroys the parity and CRN
  properties above. If the cost of respawning a generator per key ever matters, the fix is a pure,
  key-addressed cache proposed through an AMBIGUITY REPORT, not a hidden stream.
- `numpy.random.seed`, the legacy `RandomState`, or any use of the global numpy random state.
- Python's builtin `hash()` on the purpose string.
- Letting `seed_policy` enter a key, or deriving one seed from the other.
- Importing jax, scipy, pandas or torch, at module scope or inside the function.
- Adding a purpose to `PURPOSES`, or accepting a purpose not in it. The list is closed by PLAN
  section 2.15 plus `selfobs`; extending it is a spec change (CONTRACT rule 1) and needs a
  `spec/CHANGELOG.md` entry the lead writes.
- Changing the signature, the keyword-only-ness of `shape` and `dist`, or the module docstring.

## Contract rules that bite here

- **Rule 3 - STOP AND REPORT.** Choosing "the reasonable default" is a violation: do not default a
  missing distribution parameter, do not coerce an out-of-range index, do not invent a purpose name.
- **Rule 12 - WORK ORDERS (whitelist).** Read only the five entries above, write only
  `gosplan/rng.py`, run the completion command verbatim.
- **Rule 9 - RNG.** This module *is* the rule: all environment randomness goes through
  `rng.draw(seed_env, purpose, *indices)`, no direct `numpy.random` / `jax.random` under
  `gosplan/env/`, and `seed_policy` is a separate stream that never enters a key here.
- **Rule 1 - FROZEN SPEC.** The signature, the `Purpose` and `Dist` literals and the module docstring
  are the frozen interface; adding a purpose is a spec change with a CHANGELOG entry, not an
  implementer's decision.
- **Rule 2 - FROZEN TESTS.** `tests/unit/test_rng.py` is read-only; if it looks wrong, file an
  AMBIGUITY REPORT.
- **Rule 10 - MANIFEST.** Reproducibility across processes is what makes a manifest's seeds
  meaningful, which is why `zlib.crc32` and not builtin `hash`.

## Definition of done

- [ ] `tests/unit/test_rng.py` (T-U6) green.
- [ ] `uv run ruff check . && uv run ruff format --check .` clean.
- [ ] `uv run python scripts/contract_guard.py --all` clean.
- [ ] No file outside "Write only" touched.
- [ ] CI `ci-ok` green on the PR.
- [ ] The four-part report posted as the PR body: 1) tests passed/failed with output; 2) ambiguity
      reports filed, each listed, or "none"; 3) files changed - expect exactly `gosplan/rng.py`;
      4) anything you were tempted to invent and did not - in particular any parameter alias, any
      defaulting of a missing parameter, and any caching you considered.

**Comment by Xocas12** [comment]
Blocked on #52 and #53 — the `bernoulli`/`categorical` NumPy calls are pinned only in `ref/ref_step.py`, which is not on this card's exhaustive whitelist (CONTRACT rule 12), and the `selfobs` key convention contradicts `obs.py`. Both fix the same `draw` construction.
